core/templates/dev/head/pages/exploration_player/ImagePreloaderServiceSpec.js

```diff
@@ -0,0 +1,356 @@
+      var exploration = eof.createFromBackendDict(explorationDict);
+      ips.init(exploration);
+      ips.kickOffImagePreloader(exploration.getInitialState().name);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(3);
+      expect(ips.isLoadingImageFile('sIMChoice1.png')).toBe(true);
+      expect(ips.isLoadingImageFile('sIMChoice2.png')).toBe(true);
+      expect(ips.isLoadingImageFile('sIOFeedback.png')).toBe(true);
+      expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(false);
+      $httpBackend.flush(1);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(3);
+      $httpBackend.flush(1);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(2);
+      $httpBackend.flush(1);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(1);
+      $httpBackend.flush(1);
+      expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(0);
+      expect(ips.isLoadingImageFile('sIMChoice1.png')).toBe(false);
+      expect(ips.isLoadingImageFile('sIMChoice2.png')).toBe(false);
+      expect(ips.isLoadingImageFile('sIOFeedback.png')).toBe(false);
+      expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(false);
+    });
+
+  it('should properly restart pre-loading from a new state', function() {
+    var exploration = eof.createFromBackendDict(explorationDict);
+    ips.init(exploration);
+    ips.kickOffImagePreloader(exploration.getInitialState().name);
+    expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(3);
+    ips.restartImagePreloader('State 6');
+    expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(1);
+    expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(true);
+  });
+
+  it('should verify that preloader starts when state changes', function() {
+    var exploration = eof.createFromBackendDict(explorationDict);
+    ips.init(exploration);
+    ips.kickOffImagePreloader(exploration.getInitialState().name);
+    expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(3);
+    expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(false);
+    ips.onStateChange('State 6');
+    expect(ips.getFilenamesOfImageCurrentlyDownloading().length).toBe(1);
+    expect(ips.isLoadingImageFile('s6Hint1.png')).toBe(true);
+  });
+
+  it('should check that there is sync between AssetsBackendApi Service and' +
+    'ImagePreloader Service', function() {
+    var exploration = eof.createFromBackendDict(explorationDict);
+    ips.init(exploration);
+    ips.kickOffImagePreloader(exploration.getInitialState().name);
+    var filenamesOfImageCUrrentlyDownloading = (
+      ips.getFilenamesOfImageCurrentlyDownloading());
+    var imageFilesCurrentlyBeingRequested = (
+      abas.getAssetsFilesCurrentlyBeingRequested().image
+    );
+    for (x in filenamesOfImageCUrrentlyDownloading) {
+      expect(filenamesOfImageCUrrentlyDownloading[x]).toBe(
+        imageFilesCurrentlyBeingRequested[x].filename);
+    }
```

Comment:
Maybe flush one request here and after that check that the arrays are same.

Reply:
Done!